Refactor Instrument#response_set to Instrument#response_sets #3034

stgallagher opened the pull request:
### Summary

An instrument can have many response sets, yet there was code in `Instrument` for a `response_set` method that returned the first response set. I needed to find all the instances this method was being used, evaluate and change the behavior to handle multiple response sets or reference a single instance from the collection. Along the way, I pulled out some of the business logic from the controllers to the models.
##### Changes
- `ContactLinksController#setup_instrument_variables` populates instruments with reasonable defaults. Pull most of the logic out into the models. Refactored to accept multiple response sets. Refactored instances for the views
- `EventsController` Pull out the default setting logic into the model. Refactored the models to accept multiple response sets
- `Instrument` Deleted `#response_set` and `#response_set=`.  
- `Views` Adapted to multiple response sets.
- Solid chunks of testing.

stgallagher committed:
enable multiple response sets on Instrument #3034

rsutphin commented on app/controllers/contact_links_controller.rb:
```diff
@@ -97,15 +97,15 @@ def decision_page
-    @response_set  = @instrument.response_set if @instrument
-    @survey        = @response_set.survey if @response_set
+    @response_sets = @instrument.response_sets if @instrument
+    @surveys        = @response_sets.collect { |resp_set| resp_set.survey } if @response_sets
```
If you're going to pull out the surveys immediately, tell AR when you retrieve the response sets. It will be more efficient:

```
@response_sets = @instruments.response_sets.includes(:survey) if @response_set
```

Also, note whitespace difference between these two lines.

rsutphin commented on app/models/instrument.rb:
```diff
@@ -306,11 +306,15 @@ def complete?
+        break
+      end
     end
+    self.instrument_breakoff = NcsCode.for_attribute_name_and_local_code(:instrument_breakoff_code, local_code)
```
You can set `self.instrument_breakoff_code = local_code` and avoid the NcsCode lookup.

rsutphin commented on app/views/contact_links/edit.html.haml:
```diff
@@ -4,9 +4,9 @@
     = blank_safe(@person.to_s)
-    - if @response_set.complete?
+    - if @response_sets.first.complete?
       has completed
```
Should it say "has completed" if only the first RS is completed?

rsutphin commented on app/views/contact_links/edit_instrument.html.haml:
```diff
@@ -1,12 +1,12 @@
+  - if @surveys.first
     = @person
-    - if @response_set.complete?
+    - if @response_sets.first.complete?
```
Ditto here. I'm not sure on the context, but it seems like this is something that should either be repeated for each survey or be reported by consolidating results across all response sets / surveys.

rsutphin commented on spec/controllers/contact_links_controller_spec.rb:
```diff
@@ -1,6 +1,5 @@
 # -*- coding: utf-8 -*-
 
-
```
In the future, please do whitespace fixes in a separate commit. (Applies to `events_controller_spec` also.)

stgallagher committed:
corrections based on feedback #3034

- preloaded surveys when loading response sets
- set instrument_break_off_code without database
hit
- in order for 'has completed' to show up in the
view, all response_sets are enumerated and the
collection is verified complete

stgallagher commented:
Ready for Review.
